### PR TITLE
Use a modeled archival sky spectrum 

### PR DESCRIPTION
**Pull request by debora-pe**
This add the possibility to use a modeled sky spectrum (using `pypeit.wavemodel.nearIR_modelsky`) as archival sky spectrum in the spectral flexure correction.

It works well for data with wave > ~7500A, but I do not know if works better than the current method. Any thoughts?

This is currently directed to `bkg_redux_flexure`

### CODE REVIEW

File: pypeit/core/flexure.py [hunk 1]
```diff
@@ -847,25 +849,52 @@ def spec_flexure_slit_global(sciImg, waveimg, global_sky, par, slits, slitmask,
...
+                       'Not enough sky lines detected.')
+    elif obj_skyspec is not None:
+        if spec_fwhm_pix is None:
+            # measure spec_fwhm_pix
+            spec_fwhm_pix = autoid.measure_fwhm(obj_skyspec.flux.value, sigdetect=4., fwhm=4.)
+            if spec_fwhm_pix is None:
+                msgs.warn('Failed to measure the spectral FWHM using the boxcar extracted sky spectrum. '
+                          'Choose one of the provided sky files.')
+        obj_disp = np.median(np.diff(obj_skyspec.wavelength.value))
+        spec_fwhm = spec_fwhm_pix * obj_disp
```
Comment (profxj) [inline pypeit/core/flexure.py:903]: Please add a few inline docs on what is going on here

File: pypeit/core/flexure.py [hunk 2]
```diff
@@ -847,25 +849,52 @@ def spec_flexure_slit_global(sciImg, waveimg, global_sky, par, slits, slitmask,
...
-def get_archive_spectrum(sky_file):
+def get_archive_spectrum(sky_file, obj_skyspec=None, spec_fwhm_pix=None):
     """ Load an archival sky spectrum
 
     Args:
         sky_file (:obj:`str`):
-            Sky file
+            Name of the archival sky file. If equal to 'model', instead,
+            a model sky spectrum will be generated using see :func:`~pypeit.wavemodel.nearIR_modelsky`.
+        obj_skyspec (`linetools.spectra.xspectrum1d.XSpectrum1d`_, optional):
```
Comment (profxj) [inline pypeit/core/flexure.py:874]: Is it correct that this must be set if `sky_file == "model"`

if so, please note it.

File: pypeit/core/flexure.py [hunk 3]
```diff
@@ -116,17 +117,16 @@ def spat_flexure_shift(sciimg, slits, debug=False, maxlag=20):
...
     Args:
         obj_skyspec (`linetools.spectra.xspectrum1d.XSpectrum1d`_):
             Spectrum of the sky related to our object
-        arx_skyspec (`linetools.spectra.xspectrum1d.XSpectrum1d`_):
-            Archived sky spectrum
-        arx_fwhm_pix (:obj:`float`):
-            Spectral FWHM (in pixels) of the archived sky spectrum.
+        sky_file (:obj:`str`):
+            Name of the archival sky file. If equal to 'model', instead,
+            a model sky spectrum will be generated using see :func:`~pypeit.wavemodel.nearIR_modelsky`.
```
Comment (profxj) [inline pypeit/core/flexure.py]: note that it will be generated from an analysis of `obj_skyspec`

File: pypeit/core/flexure.py [hunk 4]
```diff
@@ -462,7 +467,7 @@ def spec_flex_shift_global(slit_specs, islit, sky_spectrum, arx_fwhm_pix, empty_
     flex_dict = copy.deepcopy(empty_flex_dict)
 
     # Calculate the shift
-    fdict = spec_flex_shift(slit_specs[islit], sky_spectrum, arx_fwhm_pix, mxshft=mxshft, excess_shft=excess_shft,
+    fdict = spec_flex_shift(slit_specs[islit], sky_file, mxshft=mxshft, excess_shft=excess_shft,
```
Comment (profxj) [inline pypeit/core/flexure.py:481]: as `sky_file=="model"` requires the `obj_sky_spec` this method can't use that option, right?   just checking

File: pypeit/core/flexure.py [hunk 5]
```diff
@@ -567,10 +571,10 @@ def spec_flex_shift_local(slits, slitord, specobjs, islit, sky_spectrum, arx_fwh
         msgs.info(f"Working on spectral flexure for object # {ss} in slit {slits.spat_id[islit]}")
 
         # get 1D spectrum for this object
-        obj_sky = xspectrum1d.XSpectrum1D.from_tuple((sobj.BOX_WAVE, sobj.BOX_COUNTS_SKY))
+        obj_sky = xspectrum1d.XSpectrum1D.from_tuple((sobj.BOX_WAVE[sobj.BOX_MASK], sobj.BOX_COUNTS_SKY[sobj.BOX_MASK]))
 
         # Calculate the shift
-        fdict = spec_flex_shift(obj_sky, sky_spectrum, arx_fwhm_pix, mxshft=mxshft, excess_shft=excess_shft,
+        fdict = spec_flex_shift(obj_sky, sky_file, mxshft=mxshft, excess_shft=excess_shft,
```
Comment (profxj) [inline pypeit/core/flexure.py:589]: same point as above.  seems to me `sky_file=="model"` isn't allowed here

File: pypeit/core/skysub.py [hunk 6]
```diff
@@ -1041,7 +1041,8 @@ def local_skysub_extract(sciimg, sciivar, tilts, waveimg, global_sky, thismask,
         canvas_list = points_mask + points_omask + text_mask + text_omask
         canvas.add('constructedcanvas', canvas_list)
 
-    return skyimage[thismask], nobkg_skyimage[thismask], objimage[thismask], modelivar[thismask], outmask[thismask]
+    return skyimage[thismask], nobkg_skyimage[thismask] if nobkg_skyimage is not None else None, \
```
Comment (profxj) [inline pypeit/core/skysub.py:1044]: can you document why `nobkg_skyimage` may be `None`?

Maybe add it to the `Returns` portion of the doc string

File: pypeit/par/pypeitpar.py [hunk 7]
```diff
@@ -852,8 +852,10 @@ def __init__(self, spec_method=None, spec_maxshift=None, spectrum=None,
 
         defaults['spectrum'] = 'paranal_sky.fits'
         dtypes['spectrum'] = str
-        descr['spectrum'] = 'Archive sky spectrum to be used for the flexure correction.'
-
+        descr['spectrum'] = ('Archive sky spectrum to be used for the flexure correction. '
+                             'See ``pypeit/data/sky_spec/`` for a list of available sky spectra. '
+                             'If ``model`` is used, a model sky spectrum will be generated '
```
Comment (profxj) [inline pypeit/par/pypeitpar.py:857]: point out here too it is generated from the the `objsky` spectrum, i.e. its resolution, etc.

**Review by profxj — CHANGES_REQUESTED** [review]
Mainly asking for clarity on a bit of the workflow.

Also, please add to the PR comments why this new
approach was implemented.  Thanks!

File: pypeit/extraction.py [hunk 8]
```diff
@@ -770,7 +770,7 @@ def local_skysub_extract(self, global_sky, sobjs, nobkg_global_sky=None,
             # prof_nsigma = self.par['reduce']['extraction']['std_prof_nsigma'] if IS_STANDARD else None
 
             # Local sky subtraction and extraction
-            self.skymodel[thismask], self.nobkg_skymodel[thismask], self.objmodel[thismask], self.ivarmodel[thismask], self.extractmask[thismask] \
+            self.skymodel[thismask], _tihs_nobkg_skymodel, self.ivarmodel[thismask], self.extractmask[thismask] \
```
Comment (badpandabear) [inline pypeit/extraction.py]: Should this be `_this_nobkg_skymodel` instead of `_tihs_nobkg_skymodel`

**Review by badpandabear — APPROVED** [review]
Looks good to me.

**Comment by debora-pe** [comment]
> Mainly asking for clarity on a bit of the workflow.
> 
> Also, please add to the PR comments why this new approach was implemented. Thanks!

Thanks @profxj .

This new approach was implemented to improve the flexure correction. It is not clear it is better than the current approach, but it definitely helps in the case when the object's spectral resolution is higher that the resolution of the archival spectrum.

**Review by profxj — APPROVED** [review]
very good;

merge when you wish.

thanks!

**Pull request merged**